src/utils/CatalystClientBuilder.ts

```diff
@@ -0,0 +1,148 @@
+ * N lists as the updated list.
+ */
+const REQUIRED_LISTS = 3
+export async function getUpdatedApprovedListWithoutQueryingContract({
+  preKnownServers,
+  origin,
+  requiredLists,
+  fetchApprovedCatalysts
+}: {
+  preKnownServers: { list: { address: string }[] } | { network: 'mainnet' | 'ropsten' }
+  origin: string
+  requiredLists?: number
+  fetchApprovedCatalysts?: (catalystUrl: string, origin: string) => Promise<string[] | undefined>
+}): Promise<string[] | undefined> {
+  // Set defaults if needed
+  const catalystListFetch =
+    fetchApprovedCatalysts ?? ((catalystUrl, origin) => fetchCatalystsApprovedByDAO(catalystUrl, origin))
+  const requiredAmountOfLists = requiredLists ?? REQUIRED_LISTS
+
+  // Get the list of known servers
+  const knownServers = 'list' in preKnownServers ? preKnownServers.list : CatalystsList[preKnownServers.network]
+
+  // If I don't know enough servers, then it doesn't make sense to continue
+  if (knownServers.length < requiredAmountOfLists) {
+    return undefined
+  }
+
+  // Shuffle the list
+  const shuffledPreKnownServers = shuffleArray(knownServers)
+
+  // Ask N of them for their list
+  const approvedServersList = await Promise.all(
+    shuffledPreKnownServers
+      .slice(0, requiredAmountOfLists)
+      .map((server) => server.address)
+      .map((address) => catalystListFetch(address, origin))
+  )
+
+  // Removed any failures
+  const allLists: string[][] = approvedServersList.filter(
+    (approvedServerList): approvedServerList is string[] => !!approvedServerList
+  )
+
+  // Check if we need to ask for anyone else's list
+  let i = requiredAmountOfLists
+  while (i < shuffledPreKnownServers.length && allLists.length < requiredAmountOfLists) {
+    const list = await catalystListFetch(shuffledPreKnownServers[i].address, origin)
+    if (list) {
+      allLists.push(list)
+    }
+    i++
+  }
+
+  // If I didn't manage to get al least N lists from different sources, then abort
+  if (allLists.length < requiredAmountOfLists) {
+    return undefined
+  }
+
+  // Calculate the intersection
+  const intersection = calculateIntersection(allLists)
+
+  return intersection.length > 0 ? intersection : undefined
+}
+
+function calculateIntersection(lists: string[][]): string[] {
+  const count: Map<string, number> = new Map()
+  for (const list of lists) {
+    for (const element of list) {
+      count.set(element, (count.get(element) ?? 0) + 1)
+    }
+  }
+
+  return Array.from(count.entries())
+    .filter(([_, count]) => count === lists.length)
+    .map(([element]) => element)
+}
+
+async function fetchCatalystsApprovedByDAO(catalystUrl: string, origin: string): Promise<string[] | undefined> {
+  const client = new CatalystClient(catalystUrl, origin)
+  try {
+    const servers = await client.fetchCatalystsApprovedByDAO({ timeout: '5s' })
```

Comment:
Isn't this timeout too low?

Reply:
Discussed offline: will increase timeout, but will make more calls in parallel